Heal format rpc

vadmeste opened the pull request.

krisis committed:
distribute: Make server work with multiple remote disks

This change initializes rpc servers associated with disks that are
local. It makes object layer initialization on demand, namely on the
first request to the object layer.

Also adds lock RPC service vendorized minio/dsync

krisis committed:
distribute: Make server work with multiple remote disks

This change initializes rpc servers associated with disks that are
local. It makes object layer initialization on demand, namely on the
first request to the object layer.

Also adds lock RPC service vendorized minio/dsync

fwessels committed:
Added distributed RWMutex (#2369)

fwessels committed:
Added distributed RWMutex (#2369)

krisis committed:
object layer initialization using distributed locking (#2397)

* vendorized latest minio/dsync

* wip - object layer initialization using distributed locking

krisis committed:
object layer initialization using distributed locking (#2397)

* vendorized latest minio/dsync

* wip - object layer initialization using distributed locking

krisis committed:
unify single-node and distributed namespace locking (#2401)

krisis committed:
unify single-node and distributed namespace locking (#2401)

harshavardhana committed:
Move the ObjectAPI() resource to be beginning of each handlers.

We should return back proper errors so that the clients can
retry until server has been initialized.

harshavardhana committed:
Move the ObjectAPI() resource to be beginning of each handlers.

We should return back proper errors so that the clients can
retry until server has been initialized.

harshavardhana committed:
Rename rpc-{client,server} storage-rpc-{client,server}

harshavardhana committed:
Rename rpc-{client,server} storage-rpc-{client,server}

harshavardhana committed:
Add rpc layer authentication.

harshavardhana committed:
Add rpc layer authentication.

krisis committed:
Vendorize rpc reconnect changes from minio/dsync (#2405)

krisis committed:
Vendorize rpc reconnect changes from minio/dsync (#2405)

harshavardhana committed:
Add missing rpc-client.go - missing in previous commit.

harshavardhana committed:
Add missing rpc-client.go - missing in previous commit.

harshavardhana committed:
rpc/client: Add missing rpcTokens for each rpc calls.

harshavardhana committed:
rpc/client: Add missing rpcTokens for each rpc calls.

harshavardhana committed:
Extract rpc server wrapped errors and translate to storage error.

harshavardhana committed:
Extract rpc server wrapped errors and translate to storage error.

harshavardhana committed:
rpc: Re-factor ReadFile behavior client <--> server.

Current code did not marshal/unmarshal buffers properly from
the server, The reason being buffers have to be allocated and
sent properly back to client to be consumable.

harshavardhana committed:
rpc: Re-factor ReadFile behavior client <--> server.

Current code did not marshal/unmarshal buffers properly from
the server, The reason being buffers have to be allocated and
sent properly back to client to be consumable.

krisis committed:
Implement RLock, RUnlock rpc handlers (#2437)

This would make it simplify dsync.RWMutex's algorithm to acquire a
distributed read lock and be tolerant to N/2-1 failures.

krisis committed:
Implement RLock, RUnlock rpc handlers (#2437)

This would make it simplify dsync.RWMutex's algorithm to acquire a
distributed read lock and be tolerant to N/2-1 failures.

krisis committed:
storage/rpc-client: Reconnect on network disconnect (#2436)

krisis committed:
storage/rpc-client: Reconnect on network disconnect (#2436)

awwalker committed:
splitNetPath: Add support for windows paths including volumeNames e.g ip:C:\network\path

awwalker committed:
splitNetPath: Add support for windows paths including volumeNames e.g ip:C:\network\path

harshavardhana committed:
Fix web handler initialize with distributed lazy init.

harshavardhana committed:
Fix web handler initialize with distributed lazy init.

krisis committed:
Move initGracefulShutdown before objectLayer initialization (#2468)

krisis committed:
Move initGracefulShutdown before objectLayer initialization (#2468)

krisis committed:
Lock/Rlock rpc reply was incorrect (#2479)

krisis committed:
Lock/Rlock rpc reply was incorrect (#2479)

krisis committed:
Vendorize minio/dsync for server-side read lock (#2484)

- Prevention of stale lock accumulation.
- Removal of dead code.

krisis committed:
Vendorize minio/dsync for server-side read lock (#2484)

- Prevention of stale lock accumulation.
- Removal of dead code.

harshavardhana committed:
server: Add server command line for running in distributed mode

harshavardhana committed:
server: Add server command line for running in distributed mode

harshavardhana committed:
control: Fix controller CLI handling with distributed server object layer.

Object layer initialization is done lazily fix it.

harshavardhana committed:
control: Fix controller CLI handling with distributed server object layer.

Object layer initialization is done lazily fix it.

harshavardhana committed:
rpc: client login should ignore server versions.

harshavardhana committed:
rpc: client login should ignore server versions.

krisis committed:
Layered rpc-client implementation (#2512)

krisis committed:
Layered rpc-client implementation (#2512)

vadmeste committed:
Add Heal Disk metadata (format) (cherry-picking commit eb7f181)

krisis committed:
wip: locks w/ timestamp validation (#2533)

krisis committed:
wip: locks w/ timestamp validation (#2533)

balamurugana committed:
rpc-client: remove unwanted nil check of rpcClient. (#2538)

balamurugana committed:
rpc-client: remove unwanted nil check of rpcClient. (#2538)

vadmeste committed:
Add tests for Heal Disk Metadata (cherry-picking 2129fd5)

krisis committed:
auth-rpc: Reset token on disconnect (#2542)

krisis committed:
auth-rpc: Reset token on disconnect (#2542)

harshavardhana committed:
rpc: Refactor authentication and login changes. (#2543)

- Cache login requests.
- Converge validating distributed setup.

harshavardhana committed:
rpc: Refactor authentication and login changes. (#2543)

- Cache login requests.
- Converge validating distributed setup.

vadmeste committed:
Adapt HealDiskMetadata change to distributed branch

vadmeste committed:
List only objects that need healing (#2546)

vadmeste committed:
List only objects that need healing (#2546)

harshavardhana committed:
controller/auth: Implement JWT based authorization for controller. (#2544)

Fixes #2474

harshavardhana committed:
controller/auth: Implement JWT based authorization for controller. (#2544)

Fixes #2474

harshavardhana commented on cmd/control-heal-main.go:
```diff
@@ -75,14 +75,27 @@ func healControl(ctx *cli.Context) {
 	parsedURL, err := url.Parse(ctx.Args()[0])
 	fatalIf(err, "Unable to parse URL")
 
+	client, err := rpc.DialHTTPPath("tcp", parsedURL.Host, path.Join(reservedBucket, controlPath))
```
Can you use the 'newAuthClient' style done with the latest HEAD on 'distributed' branch.

harshavardhana commented on cmd/control-heal-main.go:
```diff
@@ -75,14 +75,27 @@ func healControl(ctx *cli.Context) {
+	fatalIf(err, "Unable to connect to %s", parsedURL.Host)
+
+	// Always try to fix disk metadata
+	fmt.Print("Checking and healing disk metadata..")
```
We should come up with some form of colorized formatter here.. not in this patch though.

harshavardhana commented on cmd/controller-handlers.go:
```diff
@@ -68,6 +68,29 @@ func (c *controllerAPIHandlers) HealObject(arg *HealObjectArgs, reply *HealObjec
 }
 
+// HealDiskMetadataArgs - argument for HealDiskMetadata RPC.
+type HealDiskMetadataArgs struct {
```
If there are no arguments here re-purpose 'GenericArgs'

harshavardhana commented on cmd/controller-handlers.go:
```diff
@@ -68,6 +68,29 @@ func (c *controllerAPIHandlers) HealObject(arg *HealObjectArgs, reply *HealObjec
+}
+
+// HealDiskMetadata - heal the disk format.
+func (c *controllerAPIHandlers) HealDiskMetadata(arg *HealDiskMetadataArgs, reply *HealDiskMetadataReply) error {
```
arg should be 'args'.

harshavardhana commented on cmd/controller-handlers.go:
```diff
@@ -68,6 +68,29 @@ func (c *controllerAPIHandlers) HealObject(arg *HealObjectArgs, reply *HealObjec
+
+// HealDiskMetadata - heal the disk format.
+func (c *controllerAPIHandlers) HealDiskMetadata(arg *HealDiskMetadataArgs, reply *HealDiskMetadataReply) error {
+	objAPI := c.ObjectAPI
```
This is a function. 

```
objAPI := c.ObjectAPI()
```

harshavardhana commented on cmd/controller-handlers.go:
```diff
@@ -68,6 +68,29 @@ func (c *controllerAPIHandlers) HealObject(arg *HealObjectArgs, reply *HealObjec
+	if objAPI == nil {
+		return errInvalidArgument
+	}
+	if objAPI.HealDiskMetadata() == nil {
```
Perhaps simply 

```
if err := objAPI.HeadDiskMetadata(); err != nil {
     return err
}
reply.Success = true
return nil
```

harshavardhana commented on cmd/object-interface.go:
```diff
@@ -22,6 +22,8 @@ import "io"
 	// Storage operations.
 	Shutdown() error
+	HealDiskMetadata() error
+
```
Unnecessary new line..

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -65,6 +66,9 @@ type xlObjects struct {
 	objCacheEnabled bool
+
+	// Protection for generic metadata
+	diskMetadataLock *sync.Mutex
```
You don't need this lock, - you need to hold a lock on format.json'

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -216,6 +227,13 @@ func (xl xlObjects) Shutdown() error {
+// HealDiskMetadata function for object storage interface.
+func (xl xlObjects) HealDiskMetadata() error {
+	xl.diskMetadataLock.Lock()
+	defer xl.diskMetadataLock.Unlock()
```
We should lock 'format.json' this is a distributed setup, simultaneously invoking the same 'control' heal would corrupt 'format.json'

```
      // Acquire a distributed lock to ensure only one of the nodes
                // initializes the format.json.
                nsMutex.Lock(minioMetaBucket, formatConfigFile)
                defer nsMutex.Unlock(minioMetaBucket, formatConfigFile)
```

harshavardhana commented on cmd/control-heal-main.go:
```diff
@@ -75,14 +75,27 @@ func healControl(ctx *cli.Context) {
+	args := &HealDiskMetadataArgs{}
+	reply := &HealDiskMetadataReply{}
+	err = client.Call("Control.HealDiskMetadata", args, reply)
+	fatalIf(err, "RPC Control.HealDiskMetadata call failed")
```
Print a user friendly message here 

```
fatalIf(err, "Healing disk metadata failed.")
```

vadmeste replied:
> fatalIf(err, "Healing disk metadata failed.")
> Or "Unable to heal disk metadata"

actually, I think we need to distinguish between rpc errors (e.g. network connections) and healing api errors

vadmeste replied:
oh, I see fatalIf will print the reason of the problem besides "Unable to heal disk metadata", will do it

harshavardhana commented on cmd/control-heal-main.go:
```diff
@@ -75,14 +75,27 @@ func healControl(ctx *cli.Context) {
 	bucketName, objectName := parseBucketObject(parsedURL.Path)
 	if bucketName == "" {
-		cli.ShowCommandHelpAndExit(ctx, "heal", 1)
+		// cli.ShowCommandHelpAndExit(ctx, "heal", 1)
```
any particular reason this was removed?

vadmeste replied:
> any particular reason this was removed?

yes. I don't want to oblige the user to enter a bucket name, because there is healing format configs feature now.

harshavardhana replied:
> yes. I don't want to oblige the user to enter a bucket name, because there is healing format configs feature now.

I think we can move these type of checks into a separate function as `checkHealSyntax()`

harshavardhana committed:
server: We shouldn't exit the server in lazy init. (#2548)

Avoid fatalIf instead these are non-critical errors,
continue running the server.

 - initializing bucket notifications
 - initializing bucket policies.
 - migrating bucket policies failure.

Fixes #2547

vadmeste committed:
Merge remote-tracking branch 'upstream/distributed' into distributed_heal_format
